delta_shooter/game_mechanics.py

```diff
@@ -0,0 +1,310 @@
+        total_return += reward
+    return total_return
+
+
+def load_network(team_name: str, network_folder: Path = HERE) -> nn.Module:
+    net_path = network_folder / f"{team_name}_network.pt"
+    assert (
+        net_path.exists()
+    ), f"Network saved using TEAM_NAME='{team_name}' doesn't exist! ({net_path})"
+    model = torch.load(net_path)
+    model.eval()
+    return model
+
+
+def save_network(network: nn.Module, team_name: str) -> None:
+    assert isinstance(
+        network, nn.Module
+    ), f"train() function outputs an network type: {type(network)}"
+    assert "/" not in team_name, "Invalid TEAM_NAME. '/' are illegal in TEAM_NAME"
+    net_path = HERE / f"{team_name}_network.pt"
+    n_retries = 5
+    for attempt in range(n_retries):
+        try:
+            torch.save(network, net_path)
+            load_network(team_name)
+            return
+        except Exception:
+            if attempt == n_retries - 1:
+                raise
+
+
+def choose_move_randomly(state: np.ndarray) -> int:
+    return np.random.randint(4)
+
+
+class ShooterEnv(gym.Env):
+    def __init__(
+        self,
+        opponent_choose_move: Callable,
+        render: bool = False,
+        game_speed_multiplier: float = 1,
+    ):
+
+        self._render = render
+        self.opponent_choose_move = opponent_choose_move
+        self.game_speed_multiplier = game_speed_multiplier
+        if self._render:
+            self.init_graphics()
+        else:
+            self.screen = DummyScreen(GAME_SIZE)
+
+        self.reset()
+        self.num_envs = 1
+        self.action_space = gym.spaces.Discrete(4)
+
+        if self._render:
+            self._draw()
+
+    def reset(self) -> Tuple[np.ndarray, float, bool, Dict]:
+        self.message = ""
+        spawn_idx = list(range(len(SPAWN_POINTS)))
+        random.shuffle(spawn_idx)
+
+        player1_idx = spawn_idx.pop()
+        player2_idx = spawn_idx.pop()
+
+        self.player1 = Spaceship(
+            SPAWN_POINTS[player1_idx],
+            SPAWN_ORIENTATIONS[player1_idx],
+            player=1,
+            graphical=self._render,
+        )
+        self.player2 = Spaceship(
+            SPAWN_POINTS[player2_idx],
+            SPAWN_ORIENTATIONS[player2_idx],
+            player=2,
+            graphical=self._render,
+        )
+        self.done = False
+        self.n_actions = 0
+        return self.observation_player1, 0.0, False, {}
+
+    def init_graphics(self) -> None:
+        pygame.init()
+        pygame.display.set_caption("Space Rocks")
+        self.screen = pygame.display.set_mode(GAME_SIZE)
+        self.background = load_sprite("space", False)
+        self.clock = pygame.time.Clock()
+        self.font = pygame.font.Font(None, 64)
+
+    def _step(self, action: int, player: Spaceship) -> None:
+        """Takes a single step for one player."""
+
+        assert (
+            isinstance(action, (int, np.int64)) and 0 <= action <= 3
```

Comment:
How about `action in range(4)`? More concise and simpler to understand, no?

Reply:
my solution is O(1) and your is O(n_actions) though ;)